docs/JAX_porting_PyTorch_model.md

```diff
@@ -0,0 +1,1607 @@
+
+Let us now define all the necessary model parts in JAX using Flax module. PyTorch model's implementation can be found [here](https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html#maxvit_t), we can now list the modules and helper functions we implement in JAX:
+
+- Conv2dNormActivation
+- SqueezeExcitation
+- StochasticDepth
+- MBConv
+- RelativePositionalMultiHeadAttention
+- SwapAxes
+- WindowPartition
+- WindowDepartition
+- PartitionAttentionLayer
+- MaxVitLayer
+- MaxVitBlock
+- MaxViT
+
+Flax NNX module is very similar to PyTorch `torch.nn` module and we can simply map the following modules between PyTorch and Flax:
+- `nn.Sequential` and `nn.ModuleList` -> `nnx.Sequential`
+- `nn.Linear` -> `nnx.Linear`
+- `nn.Conv2d` -> `nnx.Conv`
+- `nn.BatchNorm2d` -> `nnx.BatchNorm`
+- Activations like `nn.ReLU` -> `lambda x: nnx.relu(x)`
+- Pooling layers like `nn.AvgPool2d` -> `lambda x: nnx.avg_pool(x, ...)`
+- `nn.AdaptiveAvgPool2d(1)` -> `lambda x: nnx.avg_pool(x, (x.shape[1], x.shape[2]))`, x is in NHWC format
+- `nn.Flatten()` -> `lambda x: x.reshape(x.shape[0], -1)`
+
+
+If PyTorch model defines a learnable parameter and a buffer:
+```python
+class Model(nn.Module):
+    def __init__(self, ...):
+        ...
+        self.p = nn.Parameter(torch.ones(10))
+        self.register_buffer("b", torch.ones(5))
+```
+equivalent code in Flax would be
+```python
+class Buffer(nnx.Variable):
+    pass
+
+
+class Model(nnx.Module):
+    def __init__(self, ...):
+        ...
+        self.p = nnx.Param(jnp.ones((10,)))
+        self.b = Buffer(jnp.ones(5))
+```
+
+To inspect NNX module's learnable parameters and buffers, we can use `nnx.state`:
+```python
+nnx_module = ...
+for k, v in nnx.state(nnx_module, nnx.Param).flat_state().items():
+    print(
+        k,
+        v.value.mean() if v.value is not None else None
+    )
+
+for k, v in nnx.state(nnx_module, (nnx.BatchStat, Buffer)).flat_state().items():
+    print(
+        k,
+        v.value.mean() if v.value.dtype == "float32" else v.value.sum()
+    )
+```
+Equivalent PyTorch code is:
+```python
+torch_module = ...
+
+for m, p in torch_module.named_parameters():
+    print(m, p.detach().mean())
+
+for m, b in torch_module.named_buffers():
+    print(
+        m,
+        b.mean() if b.dtype == torch.float32 else b.sum()
+    )
+```
+
++++
+
+Please note the some tiny differences between `torch.nn` and Flax when porting models:
+- We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
+- Convolution 2D in Flax:
+  - we need explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
+  - if `padding` should be defined as integer, e.g. 2, it should be defined as `((2, 2), (2, 2))`
+- Batch Normalization: momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
+- 4D input array in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.
```

Comment:
```suggestion
Please note there are some differences between `torch.nn` and Flax when porting models:
- We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
- For a 2d convolution:
  - In Flax, we need to explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
  - If `padding` should be defined as integer in PyTorch, e.g. 2, it should be defined as `((2, 2), (2, 2))` in Flax.
- For a batch normalization: the momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
- 4D input arrays in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.
```

Please make sure these are correct especially the padding item, the phrasing is what I am focusing on.